docs: add documentation for AWS secrets (#3295)
island-is/island.is: handbook/repository/aws-secrets.md

## Getting started

1. Configure AWS with SSO, using the following configuration:

```bash
aws configure sso

SSO start URL [None]: https://island-is.awsapps.com/start
SSO Region [None]: eu-west-1
```

Then choose the environnement of your choice. Likely to be `island-is-development01`. You be prompted for the following:

```md
CLI default client Region [eu-west-1]: <Press Enter>
CLI default output format [json]: <Press Enter>
CLI profile name [AWSPowerUserAccess-X]: <Custom name (e.g. dev)> or <Press Enter>
```

This step will add the new profile to your `~/.aws/config` file.

2. Configure AWS, using the following configuration:

```bash
aws configure

AWS Access Key ID [None]: <KEY_ID>
AWS Secret Access Key [None]: <ACCESS_KEY>
Default region name [eu-west-1]: <Press Enter>
Default output format [json]: <Press Enter>
```

You will be able to find these information on the [AWS account](https://island-is.awsapps.com/start).

Then, you will need to open the file created by AWS at `~/.aws/credentials` and add your session token to it.

```bash
[default]
aws_access_key_id = <KEY_ID>
aws_secret_access_key = <ACCESS_KEY>
aws_session_token = <SESSION_TOKEN>

> ```

Reviewer: This is redundant just to fetch secrets. If you have AWS_PROFILE set to the profile you created in #1, the secret commands should work and use your SSO profile. The SSO profile is better since it works for 8 hours while this session token only works for 1 hour.

Author: If I'm only doing `yarn get-secrets api` for example, and even if I did `aws sso login --profile dev` beforehand, it will results to `An error occurred (ExpiredTokenException) when calling the GetParametersByPath operation: The security token included in the request is expired`. I need to explicitly pass the profile for the command to pick up my logged in so profile `AWS_PROFILE=dev yarn get-secrets api`

Reviewer: Sorry, missed these replies.

Would this guide be simpler if we suggest saving the SSO profile as "default" (assuming that works), or suggest running `AWS_PROFILE=dev yarn get-secrets api`?

Voilà, you are now ready to fetch and create secrets.

{% hint style="warning" %}
Remember, the session token expires every 8 hours so you will need to update it. You can either do `export AWS_SESSION_TOKEN="SESSION_TOKEN"` or change it manually by going into `~/.aws/credentials` and replacing with the new session token generated from your AWS account.

> {% endhint %}

Reviewer: In my experience, the session token expires every 1 hour, but the SSO session lasts 8 hours. Has this been improved?

# AWS Secrets

## Prerequisites

You will need `awscli` installed

```bash

> brew install awscli

Reviewer 2: rather point to this url in order to avoid being os agnostic:
https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2.html

```bash
aws configure sso

SSO start URL [None]: https://island-is.awsapps.com/start
SSO Region [None]: eu-west-1
```

Then choose the environnement of your choice. Likely to be `island-is-development01`. You be prompted for the following:

```md
CLI default client Region [eu-west-1]: <Press Enter>
CLI default output format [json]: <Press Enter>
CLI profile name [AWSPowerUserAccess-X]: <Custom name (e.g. dev)> or <Press Enter>
```

This step will add the new profile to your `~/.aws/config` file.

2. Configure AWS, using the following configuration:

```bash
aws configure

AWS Access Key ID [None]: <KEY_ID>
AWS Secret Access Key [None]: <ACCESS_KEY>
Default region name [eu-west-1]: <Press Enter>
Default output format [json]: <Press Enter>
```

You will be able to find these information on the [AWS account](https://island-is.awsapps.com/start).

Then, you will need to open the file created by AWS at `~/.aws/credentials` and add your session token to it.

```bash
[default]
aws_access_key_id = <KEY_ID>
aws_secret_access_key = <ACCESS_KEY>
aws_session_token = <SESSION_TOKEN>
```

Voilà, you are now ready to fetch and create secrets.

{% hint style="warning" %}

> Remember, the session token expires every 8 hours so you will need to update `~/.aws/credentials` with the new session token generated on your AWS account.

Reviewer 2: Much easier to just export it as environment variables 🙈

Author: Good point, I forgot there were many ways to do that..

{% endhint %}

## Usage to fetch secrets

> You should now be able to fetch secrets for the project you need. You can verify it by opening the `.env.secret` file at the root.

Reviewer 2: Maybe recommend `dotenv` as well?

Reviewer 2: Whoops... meant `direnv` to automatically export these environment variables

```bash
yarn get-secrets <project> [options]
```

**Example**:

```bash
yarn get-secrets api --reset
```

If you configured multiples profiles using `aws configure sso` you can choose which profile to run to get the secrets:

```bash
AWS_PROFILE=dev yarn get-secrets api
```

You can find your profile listed in `~/.aws/config`, e.g.

```md
[default]
region = eu-west-1
output = json
[profile dev]
...
[profile staging]
...
```

## Usage to create secrets

You can run the following command and will be prompted for input.

```bash
yarn create-secret
```

> You will be asked for a _secret name_ that will be added to the `/k8s/` secrets namespace and a _secret value_. The length of the _secret name_ should be from 6-32 characters long.

Reviewer 2: can be 128 chars long

{% hint style="warning" %}
Only alphanumeric characters, `/` and `-` are allowed.
{% endhint %}

### Making dev secrets available locally

Environment variables that should not be tracked but needed locally should be added to the `.env.secret` file.

Additionally, if that same variable is also stored in AWS Parameter Store, the secret can be labeled with the `dev` label from `History` -> `Attach labels`.

> All secrets labeled with the `dev` label can be fetched using `yarn get-secrets`.

Reviewer 2: `yarn get-secrets <your-project>`

### Environment variables with static websites

> To be able to access environment variables in purely static projects, you need

Reviewer 2: This is kind of not related to aws. Should be documented elsewhere and referenced here.

- You will also need access to the AWS account. Ask someone from DevOps to send you an invitation.

## Getting started

### Using AWS SSO

Using SSO is the most straight forward solution. You won't need to go by yourself on your AWS account and it will open the needed url for you.

1. Run the sso command for the first time

```bash
aws configure sso

SSO start URL [None]: https://island-is.awsapps.com/start
SSO Region [None]: eu-west-1
```

Then choose the environnement of your choice. Likely to be `island-is-development01`. You will be prompted for the following:

```md
CLI default client Region [eu-west-1]: <Press Enter>
CLI default output format [json]: <Press Enter>
CLI profile name [AWSPowerUserAccess-X]: <Custom name (e.g. dev)> or <Press Enter>
```

This step will add the new profile to your `~/.aws/config` file. If you choose `dev` as profile's name, you will see `[profile dev]` in there.

2. Ready to use

You can now pass your profile to the scripts we have.

```bash
AWS_PROFILE=<profile-name> yarn get-secrets <project-name> # e.g. profile-name -> dev as seen above
```

3. Refresh your profile

> Once you configured the previous profile, you can run the following command to renew your SSO credentials.

Reviewer: ```suggestion
The SSO credentials only lasts 8 hours, after which AWS commands start failing. You can run the following command to renew your SSO credentials.
```

```bash
aws configure sso --profile <profile-name> # e.g. profile-name -> dev as seen above
```

It will open the browser, go to your AWS account to log in and will refresh your credentials and you are ready to use the AWS commands again.

> 
> {% hint style="warning" %}
> The SSO login expires every 8 hours.
> {% endhint %}

Reviewer: ```suggestion
```

### Using AWS session

This method is more manual where you will need to export environments or change file by yourself.

You will need to go to your [AWS account](https://island-is.awsapps.com/start) and get the required credentials for the account you need.

1. Option 1: Set environments variables

> You can first copy/paste theses environments variables on your terminal.

Reviewer: ```suggestion
You can copy/paste these environment variables to your terminal:
```

```bash
export AWS_ACCESS_KEY_ID=AWS_ACCESS_KEY_ID_EXAMPLE
export AWS_SECRET_ACCESS_KEY=AWS_SECRET_ACCESS_KEY_EXAMPLE
export AWS_SESSION_TOKEN=AWS_SESSION_TOKEN_EXAMPLE
```

2. Option 2: Edit `~/.aws/credentials`

Or copy/paste the values in the `~/.aws/credentials` file.

```bash

> [X_AWSPowerUserAccess]

Reviewer: ```suggestion
[default]
```
?

Author: Correct, that's a tricky one, AWS website gives a custom profile name instead of default

aws_access_key_id = <KEY_ID>
aws_secret_access_key = <ACCESS_KEY>
aws_session_token = <SESSION_TOKEN>
```

3. Refresh your profile

> You will need to log in to your AWS account and get the refreshed credentials, with one of the following method above.

Reviewer: Maybe add this after "Ready to use" and explain similarly as above:

```
The session token only lasts 1 hour, after which AWS commands start failing. You will need to log in to your AWS account and get new credentials, with one of the above methods.
```

**With SSO**

```bash
AWS_PROFILE=<profile-name> yarn get-secrets <project> [options]
```

**Without SSO**

```bash
yarn get-secrets <project> [options]
```

**Example**:

```bash
yarn get-secrets api
```

You can verify it by opening the `.env.secret` file at the root, or inside your code using for example:

```typescript
const { MY_SECRET_KEY } = process.env
```

You can also add the `--reset` argument to the command, that will reset the .env.secret file.

### Troubleshoot

If you get the following error message, you will need to refresh your credentials as explained above.

```bash
An error occurred (ExpiredTokenException) when calling the GetParametersByPath operation: The security token included in the request is expired
```

## Usage to create secrets

You can run the following command and will be prompted for input.

```bash
yarn create-secret
```

You will be asked for a _secret name_ that will be added to the `/k8s/` secrets namespace, a _secret value_ and the _secret type_ (`SecureString` or `String`).

### Example

```bash
➜ yarn create-secret

Secret name: /k8s/my-app/MY_APP_KEY
# Name: Ok!
# Length: Ok!

Secret value: a-very-secure-secret

> # Length: Ok!

Reviewer: I haven't used this before, isn't there special settings to specify dev/staging/prod secrets?

Author: Good point, @busla any input about this?

Reviewer 3: Apologies!

This mention totally passed my radar.

This would be optimal but the environment is bound to the users SSO session so we would need to make sure the role on the user has cross-account access. I'll discuss this with the team.

SecureString [Y/n]? # [enter] for SecureString
# SecureString selected

Add tags? [y/N]? # [enter] to skip creating tags

> Example: Key=Foo,Value=Bar Key=Another,Value=Tag: # note: Key and Value are case sensitive! Create multiple tags by separating with whitespace.

Reviewer: Would be nice if this script asked if you want to add the dev label (separate from tags) to allow devs to fetch the dev secret (should add this to the backlog).
